File: cpp/matrix.cpp
#include "matrix.h"

Matrix::Matrix()
{

}

QVector<QVector<float>> Matrix::dot(QVector<QVector<float> > mat_a, QVector<QVector<float> > mat_b)
{
    QVector<QVector<float>> res(mat_a.length());

    if (mat_a[0].length() != mat_b.length()) {
        qDebug() << "matrix.cpp: The columns of the first matrix must be equal to the rows of the second matrix";
        return res;
    }

    /*qDebug() << "matrix.cpp: mat_a:" << mat_a;
    qDebug() << "matrix.cpp: mat_b:" << mat_b;*/

    for (int index_rowA = 0; index_rowA < mat_a.length(); index_rowA++) {
        QVector<float> res_row(mat_b[0].length());
        QVector<float> rowA = mat_a[index_rowA];

        for (int index_colB = 0; index_colB < mat_b[0].length(); index_colB++) {
            float summation = 0;

            for (int index = 0; index < rowA.length(); index++) {
                summation += (rowA[index] * mat_b[index][index_colB]);
            }

            res_row[index_colB] = summation;
        }

        res[index_rowA] = res_row;
    }

    return res;
}

QVector<QVector<float> > Matrix::add(QVector<QVector<float> > mat_a, QVector<QVector<float> > mat_b)
{
    QVector<QVector<float>> res(mat_a.length());

    for (int index_rowA = 0; index_rowA < mat_a.length(); index_rowA++) {
        QVector<float> res_row(mat_b[0].length()); // could be mat_a[0].length as they are of equal dimensions
        QVector<float> rowA = mat_a[index_rowA];

        for (int index = 0; index < rowA.length(); index++) {
            res_row[index] = rowA[index] + mat_b[index_rowA][index];
        }

        res[index_rowA] = res_row;
    }

    return res;
}

QVector<QVector<float> > Matrix::subtract(QVector<QVector<float> > mat_a, QVector<QVector<float> > mat_b)
{
    QVector<QVector<float>> res(mat_a.length());

    for (int index_rowA = 0; index_rowA < mat_a.length(); index_rowA++) {
        QVector<float> res_row(mat_b[0].length()); // could be mat_a[0].length as they are of equal dimensions
        QVector<float> rowA = mat_a[index_rowA];

        for (int index = 0; index < rowA.length(); index++) {
            res_row[index] = rowA[index] - mat_b[index_rowA][index];
        }

        res[index_rowA] = res_row;
    }

    return res;
}

QVector<QVector<float> > Matrix::multiply(QVector<QVector<float> > mat_a, QVector<QVector<float> > mat_b)
{
    QVector<QVector<float>> res(mat_a.length());

    for (int index_rowA = 0; index_rowA < mat_a.length(); index_rowA++) {
        QVector<float> res_row(mat_b[0].length()); // could be mat_a[0].length as they are of equal dimensions
        QVector<float> rowA = mat_a[index_rowA];

        for (int index = 0; index < rowA.length(); index++) {
            res_row[index] = rowA[index] * mat_b[index_rowA][index];
        }

        res[index_rowA] = res_row;
    }

    return res;
}

QVector<QVector<float> > Matrix::transpose(QVector<QVector<float> > mat)
{
    QVector<QVector<float>> transposed(mat[0].length());

    for (int index = 0; index < transposed.length(); index++) {
        QVector<float> row(mat.length());
        transposed[index] = row;
    }

    for (int rowIndex = 0; rowIndex < mat.length(); rowIndex++) {
        QVector<float> row = mat[rowIndex];

        for (int index = 0; index < row.length(); index++) {
            transposed[index][rowIndex] = row[index];
        }
    }

    return transposed;
}

File: prototypes/Scripts/Matrix.js
function dot(mat_a, mat_b) {
    /*
     * 1D matrix must be [array] not array
     * Need columns of first matrix to be equal to the rows of second matrix
     * Use transpose if there is only one row (array) in the second matrix
     */

    if (mat_a[0].length !== mat_b.length) {
        console.log("The columns of the first matrix must be equal to the rows of the second matrix")
        return []
    }

    var res = []
    mat_a.forEach(function(rowA, index_rowA) {
        var res_row = []
        for (var index_colB = 0; index_colB < mat_b[0].length; index_colB++) {
            var summation = 0
            rowA.forEach(function(element, index) {
                summation += (element * mat_b[index][index_colB])
            })
            res_row.push(summation)
        }
        res.push(res_row)
    })

    return res
}

function add(mat_a, mat_b) {
    var res = []
    mat_a.forEach(function(rowA, index_rowA) {
        var res_row = []
        rowA.forEach(function(element, index) {
            res_row.push(element + mat_b[index_rowA][index])
        })
        res.push(res_row)
    })

    return res
}

function subtract(mat_a, mat_b) {
    var res = []
    mat_a.forEach(function(rowA, index_rowA) {
        var res_row = []
        rowA.forEach(function(element, index) {
            res_row.push(element - mat_b[index_rowA][index])
        })
        res.push(res_row)
    })

    return res
}

function multiply(mat_a, mat_b) {
    var res = []
    mat_a.forEach(function(rowA, index_rowA) {
        var res_row = []
        rowA.forEach(function(element, index) {
            res_row.push(element * mat_b[index_rowA][index])
        })
        res.push(res_row)
    })

    return res
}

function transpose(mat) {
    var transposed = []
    for (var index = 0; index < mat[0].length; index++)
        transposed.push([])

    mat.forEach(function(row, rowIndex) {
        row.forEach(function(element, index) {
            transposed[index][rowIndex] = element
        })
    })

    return transposed
}

function fillCheck(mat) {
    /*
     * Checkes whether a matrix is filled to a square/rectangle shape
     * [[1, 2], [3, 4]] is filled (square)
     * [[1, 2], [3, 4], [5, 6]] is filled (rectangle)
     * [[1, 2], [3, 4, 5]] is not filled
     */
    var comareCount = mat[0].length
    if (mat.length > 0) {
        for (var index = 1; index < mat.length; index++)
            if (mat[index].length !== comareCount) return false
    }
    return true
}

function sizeCheck(mat_a, mat_b) {
    if (mat_a.length !== mat_b.length) return false
    if (mat_a[0].length !== mat_b[0].length) return false
    return true
}
